File: CPP/ByRound/1940/1940_B_Three_Arrays.cpp
// Problem: CF 1940 B - Three Arrays
// https://codeforces.com/contest/1940/problem/B

/*
Purpose: This code solves the "Three Arrays" problem where we need to maximize the sum of the final elements of two arrays A and B, constructed according to specific rules involving three input arrays D, L, and R, and initial values a and b.

Approach:
- Dynamic Programming with greedy optimization using sorting.
- The key idea is to decide at each step which operation (apply min to A or B) leads to a better final result.
- We process the input arrays in reverse order, accumulating sums in D.
- For each element, we compute adjusted values of L and R by adding the accumulated D values.
- Using a sorting technique to organize indices based on the adjusted L values.
- Then we iterate through the sorted indices to simulate choices and compute optimal sums.

Time Complexity: O(n log n) - due to the sorting step. The rest of the processing is linear.
Space Complexity: O(n) - for storing arrays and indices.

Algorithms/Techniques:
- Dynamic Programming
- Sorting with custom comparator
- Greedy choice strategy
*/

#include <stdio.h>
#define N 100000
#define INF 0x3f3f3f3f3f3f3f3fLL
long long min(long long a, long long b) { return a < b ? a : b; }
long long max(long long a, long long b) { return a > b ? a : b; }
unsigned int X = 12345;
int rand_() { return (X *= 3) >> 1; }
long long dd[N + 1], xx[N + 1], yy[N + 1];

// Custom quicksort implementation for sorting indices
void sort(int *ii, int l, int r) {
  while (l < r) {
    int i = l, j = l, k = r, i_ = ii[l + rand_() % (r - l)], tmp;
    while (j < k)
      if (xx[ii[j]] == xx[i_])
        j++;
      else if (xx[ii[j]] < xx[i_]) {
        tmp = ii[i], ii[i] = ii[j], ii[j] = tmp;
        i++, j++;
      } else {
        k--;
        tmp = ii[j], ii[j] = ii[k], ii[k] = tmp;
      }
    sort(ii, l, i);
    l = k;
  }
}

int main() {
  static int ii[N]; // array to store indices
  int n, i;
  long long x, y, ans;

  // Read input
  scanf("%d", &n);
  for (i = 0; i < n; i++)
    scanf("%lld", &dd[i]);

  // Accumulate D from right to left
  for (i = n - 1; i >= 0; i--)
    dd[i] += dd[i + 1];

  // Compute adjusted L values and store in xx
  for (i = 1; i <= n; i++) {
    scanf("%lld", &xx[i]);
    xx[i] += dd[i];  // Add accumulated sum to L
  }

  // Compute adjusted R values and store in yy
  for (i = 1; i <= n; i++) {
    scanf("%lld", &yy[i]);
    yy[i] += dd[i];  // Add accumulated sum to R
  }

  // Add the initial a and b to x and y respectively after adding accumulated sum
  scanf("%lld%lld", &x, &y), x += dd[0], y += dd[0];

  // Initialize index array
  for (i = 1; i <= n; i++)
    ii[i] = i;

  // Sort indices based on adjusted L values
  sort(ii, 1, n + 1);

  ans = 0; // Initialize answer

  // Iterate and calculate maximum value using greedy strategy
  for (i = 1; i <= n + 1; i++) {
    // At each step, try to find the best possible combination
    ans = max(ans, min(x, i <= n ? xx[ii[i]] : INF) + y);
    if (i <= n)
      y = min(y, yy[ii[i]]);  // Update y using smallest R value so far
  }

  // Output the maximum sum
  printf("%lld\n", ans);
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/B/1940_B_Three_Arrays.py
# Problem: CF 1940 B - Three Arrays
# https://codeforces.com/contest/1940/problem/B

"""
Algorithm: Dynamic Programming with Sorting and Prefix Minimization
Time Complexity: O(n log n) due to sorting
Space Complexity: O(n) for storing transformed arrays and auxiliary structures

This problem involves constructing two arrays A and B from given input arrays D, L, R, and initial values a0, b0.
At each step i, we add D[i] to both A[i] and B[i], and then choose to apply either:
- A[i] = min(A[i], L[i]) 
- B[i] = min(B[i], R[i])

We want to maximize A[n] + B[n]. The approach uses dynamic programming with reversed preprocessing and greedy selection.

Key observations:
1. We process indices in reverse order (from n to 1) to compute prefix sums
2. For each index i, we transform L[i] and R[i] by adding prefix sum up to i+1 
3. After transforming, we sort pairs (L[i], R[i]) to consider optimal choices greedily
4. Finally, we calculate maximum sum by evaluating all viable combinations of min operations

The solution works by:
- Preprocessing to compute cumulative sums and transformed arrays
- Sorting by L[i] values (first element of pairs)
- Using prefix minimums to find best B values and evaluating valid transitions
"""

import math
import sys

ints, dbg, yes = (
    lambda: list(map(int, input().split())),
    lambda *args, **kwargs: print(*args, file=sys.stderr, **kwargs),
    lambda x: print("YES" if x else "NO"),
)

(tests,) = (1,)

MOD = 10**9 + 7
for test in range(tests):
    (n,) = ints()
    d, l, r = [0] + ints(), [0] + ints(), [0] + ints()  # Add dummy zero at start for 1-indexing
    a0, b0 = ints()
    csum = 0
    goida = []  # Store (L[i], R[i]) pairs after transformation
    for i in range(n, 0, -1):  # Reverse iteration through indices from n down to 1
        # Add cumulative sum of D elements to L and R for current index
        l[i] += csum
        r[i] += csum
        goida.append((l[i], r[i]))
        csum += d[i]
    a0 += csum  # Final contribution to A[0] from all D values
    b0 += csum  # Final contribution to B[0] from all D values
    ans = 0
    to_make = math.inf  # Tracks minimum B value possible so far (greedy prefix)
    goida.sort()  # Sort by first element of pairs (L[i] values)
    ans = max(ans, goida[0][0] + b0)  # Initial case when choosing first L value
    for i in range(n):
        # Update prefix minimum for B values
        to_make = min(to_make, goida[i][1], b0)
        # If next L value is different, check transition to it (greedy)
        if i != n - 1 and goida[i + 1][0] != goida[i][0]:
            ans = max(ans, min(a0, goida[i + 1][0]) + to_make)
    ans = max(ans, a0 + min(r[1:]))  # Consider final case without any L transition
    print(ans)


# https://github.com/VaHiX/CodeForces/